### PR TITLE
Example of optional Path extractor added

### PR DESCRIPTION
**Pull request by sukhmel**
An example of how a route handler with an `Option<Path<_>>` should be used is added. This should allow for a better understanding of how to idiomatically use multiple paths with the same handler, and allow new users not to stumble upon rejections for a seemingly optional parameter missing

Refs: #1662

## Motivation

For me as a new Axum user, the documentation was not clear as to how to define an optional path parameter.  After creating an issue and seeing a response from maintainers on how this should have been done, it became apparent that having a path parameter extractor optional is not enough as we'd need to somehow guess what the route would look like in such a case. Nevertheless, I find it possible that there will be others like me, who will lack the necessary insight and be puzzled by rejected requests.

I assume that this could be mitigated by stating the fact that Path extractors are a special case of optional extractors, and new users could benefit from an example of how those should be used. 

## Solution

A basic example of how a handler with an optional path parameter extractor could be used for handling several routes

### CODE REVIEW

**Review by davidpdrsn — COMMENTED** [review]
Thanks! Though I think it should be moved to the `Path` extractor, since its specific to that.

**Comment by jplatte** [comment]
For a `String` path parameter it isn't really relevant, but `Path` extraction can fail in another way than the route not having the required parameter(s): argument deserialization (first variant of [`PathRejection`](https://docs.rs/axum/latest/axum/extract/rejection/enum.PathRejection.html)). Because of that, I'm not sure whether we should encourage this pattern in general.

**Comment by davidpdrsn** [comment]
Hm yeah I was thinking the same. I thought it was more relevant to middleware maybe.

**Comment by sukhmel** [comment]
@jplatte as for deserialization errors, I had something like this in mind, since we have a numeric IDs and a special case of non-numeric ID to represent the current user. So having it fail and then try another route with that special case is exactly what is needed. As for a String parameter in an example, maybe I should replace it with something that must be parsed like u128 instead?

@davidpdrsn maybe it's better to move the example to Path, but still a hint in the Optional extractors section is necessary, I think. I for one would not try searching for way to make a specific extractor optional in it's doc, if there exist a dedicated doc section about the general case of optional extractors.

**Comment by jplatte** [comment]
> @jplatte as for deserialization errors, I had something like this in mind, since we have a numeric IDs and a special case of non-numeric ID to represent the current user. So having it fail and then try another route with that special case is exactly what is needed.

You are misunderstanding. axum never tries another route when an extractor fails. Instead, for two routes for `/user/me` and `/user/:user_id`, the former is determined to be more specific and its handler will be called on that exact path. However, your handler will still be called for `/user/not_an_integer` which also matches `/user/:user_id`, and because `Path<IntegerType>` extraction will fail with a deserialization error, your `Option<Path<IntegerType>>` extractor parameter will be `None`.

**Comment by sukhmel** [comment]
thanks for the explanation, that will certainly help me when defining routes

regarding the concern of encouraging this pattern, I find it possible that the user could wrap it in Result instead, to be able to check whether this was a missing parameter or a parameter, that failed to parse. Maybe I could rewrite the example to reflect that?

**Comment by jplatte** [comment]
Maybe the new docs could be on the `Path` page but linked to from the extract module page?

**Comment by sukhmel** [comment]
@jplatte ok, I'll try to device a way to phrase and link it that way

**Comment by sukhmel** [comment]
not sure if I need to tag anyone after I've committed changes, please do tell me if I should, as I am not used to this workflow

I have moved an example and added a link, not sure if it is worded in a proper way, but it seems ready for a review

File: axum/src/extract/path/mod.rs [hunk 1]
```diff
@@ -143,6 +143,52 @@ use std::{
...
+/// }
+///
+/// let app = Router::new()
+///     .route("/users/:id", get(get_user))
+///     .route("/users", get(get_user));
+///
+/// # async {
+/// # axum::Server::bind(&"".parse().unwrap()).serve(app.into_make_service()).await.unwrap();
+/// # };
+/// ```
```
Comment (davidpdrsn) [inline axum/src/extract/path/mod.rs:189]: I'm not sure we should show examples of code we _don't_ recommend users write 🤔

I think this is a nicer way of doing it

```rust
async fn with_params(Path(id): Path<u32>) -> Response {
    do_thing(Some(id)).await
}

async fn without_params() -> Response {
    do_thing(None).await
}

async fn do_thing(param: Option<u32>) -> Response { ... }

Router::new()
    .route("/:id", get(with_params))
    .route("/", get(without_params));
```

I'd be okay with including something like that in the docs for `Path`.

File: axum/src/docs/extract.md [hunk 2]
```diff
@@ -251,6 +251,9 @@ let app = Router::new().route("/users", post(create_user));
 # };
 ```
 
+`Path` extractors have a specific behaviour and should only be used with `Result`
+wrapper, see [`path`](`Path`#optional-path-parameters) module doc for details.
```
Comment (davidpdrsn) [inline axum/src/docs/extract.md:255]: Hm I don't think I'd say that `PathRejection` is special. All rejections work like this. For example `Option<Json<T>>` doesn't tell you whether the `content-type: application/json` was missing or whether there was a deserialization error.

I think I'd just leave out this paragraph.

File: axum/src/extract/path/mod.rs (hunk 1 above)
Comment (sukhmel) [reply axum/src/extract/path/mod.rs:189]: I thought the point was not to use the `Option<Path>` rather than not to use `Result<Path>`. Maybe it would be OK to add both this to the example and `Result<Path>` as a manual handling example?

File: axum/src/docs/extract.md (hunk 2 above)
Comment (sukhmel) [reply axum/src/docs/extract.md:255]: Well, what I meant is that a missing path parameter will be rejected early and not reach the handler at all. Maybe I can rephrase this to emphasize that this is about rejection based on a route before any handler is called

File: axum/src/extract/path/mod.rs (hunk 1 above)
Comment (davidpdrsn) [reply axum/src/extract/path/mod.rs:189]: I don't think `Result<Path>` (or `Option<Path>`) should be used to implement optional path params. So I don't think we should have an example that shows that. `Result<Path>` can be used to customize deserialization errors but thats already covered by the existing extractor docs.

File: axum/src/docs/extract.md (hunk 2 above)
Comment (davidpdrsn) [reply axum/src/docs/extract.md:255]: But that is also not special for `Path`. Any rejection will cause the handler to not be called.

Comment (sukhmel) [reply axum/src/docs/extract.md:255]: I seem to misunderstand how this works, but I can't get a clear picture. Let's take the optional extractor example, there we have `Option<Json<Value>>` and missing json does not mean that we will get a rejection without calling the handler.

In the case of Path extractor, on the other hand, the parameters should match the route path first and only then they are passed to handler, that might also reject the result based on its own parameters.

File: axum/src/extract/path/mod.rs (hunk 1 above)
Comment (sukhmel) [reply axum/src/extract/path/mod.rs:189]: I see, it makes sense to rewrite this example then, will do

File: axum/src/docs/extract.md (hunk 2 above)
Comment (davidpdrsn) [reply axum/src/docs/extract.md:255]: > In the case of Path extractor, on the other hand, the parameters should match the route path first and only then they are passed to handler, that might also reject the result based on its own parameters.

Thats not correct. If you do `.route("/", get(|params: Option<Path<String>>| async {}))`, i.e a route without any path params using `Option<Path<_>>` then the handler is still always called. `params` will always be `None`.

A mismatch between the route and `Path` triggers a rejection which `Option` discards and yields a `None` (https://github.com/tokio-rs/axum/blob/main/axum-core/src/extract/mod.rs#L153).

Comment (sukhmel) [reply axum/src/docs/extract.md:255]: So `Option<Json<T>>` being None is not a rejection, or is that something else I am missing?

Comment (davidpdrsn) [reply axum/src/docs/extract.md:255]: What do you mean by " is not a rejection"? `Option<Json<_>>` is an extractor whose rejection is `Infallible` (ie never rejects)

Comment (davidpdrsn) [reply axum/src/docs/extract.md:255]: I feel like you're misunderstanding what I'm trying to say.

> if I do `.route("/", get(|params: Option<Path<String>>| async {}))` it will handle the missing path parameter as expected, because there is no parameter in the path, nowhere to get it from

In this case the handler will _always_ be called and `params` will be `None`. No rejections will happen.

> if I do `.route("/:parameter/suffix", get(|params: Option<Path<String>>| async {}))` then there is no way to get this handler called

Here the handler will also _always_ be called and `params` will be a `Some`. Again no rejections. If the route matches then the path parameter can be extracted and `String` cannot fail deserialization.

> even if I do `.route("/:parameter", get(|params: Option<Path<String>>| async {}))` it will end up rejected both without trailing slash, and with it (because with trailing slash there will be a redirection ending with rejection, too)

Same here as the previous example. It'll always work.

**Comment by davidpdrsn** [comment]
I'm sorry but I think I'll close this. I worry that adding this to the docs is just gonna confused people and optional path parameters is probably not something we wanna encourage anyway. Feel free to ask follow up questions though!

File: axum/src/docs/extract.md (hunk 2 above)
Comment (sukhmel) [reply axum/src/docs/extract.md:255]: I was talking about those routes being called without parameter, like `//suffix`

well, anyway, maybe this issue will at least pop up in search for someone. This week I've already explained to a colleague why they shouldn't use `Option<Path>` directly in axum handler and rather create a shim like you showed, so this is not something that's not going to be used just because it is not clearly mentioned in the docs.

I'd argue that it would be better to change the relevant part as per your comments to show the way it should be used rather than have everyone come up with their own ideas of what to do in such a case
